### PR TITLE
Add optims to detect when SoA matrices can be used

### PR DESCRIPTION
**Pull request by SteveBronder**
Ignore for now, I'm debugging a few things and will ping @rybern and @rok-cesnovar when this is ready for review / will fill out the top level comment describing the scheme and changes here.

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation PR is here: <LINK>
    - [x] OR, no user-facing changes were made

## Release notes

Allows `log_prob_impl` to use Struct of Arrays type for reverse mode autodiff

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

**Comment by SteveBronder** [comment]
Yeeessss first pass, I need to fixup one thing in the math library but then I'll clean this up and get it ready for review!

**Comment by wds15** [comment]
Let me know when its time to run test models…

**Comment by SteveBronder** [comment]
Alrighty I think this is ready to start reviewing! This is pending a math PR coming in but I setup jenkins to use that branch of math until that math branch is merged to develop.

@rok-cesnovar can you take a look at the code in `stan_math_backend`? Also is there a way to get a binary for @wds15 to try some stuff out?

@rybern if you can start looking at the optim code that would be super

Close! Very excited!

**Comment by rok-cesnovar** [comment]
> @rok-cesnovar can you take a look at the code in stan_math_backend? 

Those changes look good!

> Also is there a way to get a binary for @wds15 to try some stuff out?

You can use the Github Action for that (https://github.com/stan-dev/stanc3/actions/workflows/binaries.yml) but as this PR is done on your fork, you have to run that on your fork.

File: src/analysis_and_optimization/Optimize.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Optimize.ml]: Related to our call I have an example where setting `allow_kill` to `false` changes the ad level optimizer. On the current commit (`git checkout 3e32a62` from my fork) change line 1144 in `minimal_variables_mfp` of the monotone framework to 

```
  let (module Transfer3) =
    minimal_variables_fwd2_transfer true flowgraph_to_mir
      (Map.map ~f:(fun x -> x.entry) rev_min_vars_mfp)
```

then run `dune runtest` and you'll see it flips more stuff to double. I'm looking right now into whether theres any bugs in the SoA optim code that could cause this but I think there is something a bit odd about the effect here on the ad level optimizer

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:324]: I just updated the docs with all of the logic rules. This one is if the lhs is a matrix the rhs cannot be a mix of data matrices and parameter scalars. The problem is that in the math library we always return a `Matrix<var>` (the AoS matrix type) for `f(data matrix, param scalar)` like non-unary functions. I'm gonna rewrite this section to lay out whats going on more. There is another intricate thing here I found where we can actually keep things SoA for instances like 

```stan 
functions {
    matrix okay_reduction(real sum_x, matrix y) {
        return sum_x + y;
    }
}
parameters {
    matrix [5, 10] soa_x;
    matrix [5, 10] aos_x;
}
transformed parameters {
    matrix[5, 10] tp_matrix_from_udf_reduced_soa = okay_reduction(sum(soa_x), aos_x);
}
```

Since `soa_x` is only the sum of `soa_x` is used in the UDF we don't need to put it on the list of parameters that need demoted. So that makes the logic a little more complicated but I'll write all that up

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:331]: Oh that's not right! Fixed this up, I think I changed it as I was trying to debug something.

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:306]: Oh! That was def wrong I fixed the naming and logic here

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:366]: This is a really good point. Let me go back and check on this. I think this is going to require something a bit more clever

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:405]: So I _think_ this is fine because of how `SoA` only effects matrices. So for `size(A)` it does return an `int`, but for the purpose of deduction of what is allowable to be `SoA` that function returning back a non-matrix type is okay. Another example is like `array[] exp_quad_cov(vector, vector)`. So if we saw something like 

```stan
parameters {
  vector[N] Y;
  vector[N] Z;
}
transformed parameters {
  vector[N] X = to_vector(cov_exp_quad(Y, Z));
}
```

Here since `cov_exp_quad(Y, Z)` returns an array of reals we already know it's too expensive to convert that array of reals (structs) to SoA.

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:419]: I think we also need to check both are empty since the empty set is a subset of the empty set, no?

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:108]: Yes!

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:110]: Yes!

**Comment by SteveBronder** [comment]
Alrighty @rybern I think this is ready for another review! I added everything from the last review (I think) and added an option for the compile flags to turn off the optim if we need

**Comment by rybern** [comment]
Awesome! I'll try to hit this by the end of the week if that's okay. Did you have a chance to ask Matthijs about the ad-level optimization issue you found?

**Comment by SteveBronder** [comment]
@rybern all good! I got totally distracted and need to email him today.

**Comment by SteveBronder** [comment]
@VMatthijs @rybern so the good news is that removed the rev and extra forward pass everything looks good! The bad news is that when I tried making the flowgraph circular it spits out bad code for the [deep_dependence.stan](https://github.com/SteveBronder/stanc3/blob/feature/soa-optim/test/integration/good/mem_patterns/deep_dependence.stan) example.

I'm not really using it here but I added the [`make_circular_flowgraph`](https://github.com/SteveBronder/stanc3/blob/feature/soa-optim/src/analysis_and_optimization/Monotone_framework.ml#L137) function and then use it [here](https://github.com/SteveBronder/stanc3/blob/feature/soa-optim/src/analysis_and_optimization/Optimize.ml#L1068). You can see it's result by having [`minimal_variables_mfp`](https://github.com/SteveBronder/stanc3/blob/feature/soa-optim/src/analysis_and_optimization/Monotone_framework.ml#L1061) just return back `fwd1_min_vars_mfp` instead of doing the loop check.

Looking at the [`mfp`](https://github.com/SteveBronder/stanc3/blob/feature/soa-optim/src/analysis_and_optimization/Monotone_framework.ml#L826) function I don't really see how adding those initials to the exits would force it to continuously loop as that function just iterates over `F.successors` once then does a while loop popping the `workstack` till it's size zero. Or am I missing something?

**Comment by rybern** [comment]
> so the good news is that removed the rev and extra forward pass everything looks good!
Yay! That's a good fallback

> The bad news is that when I tried making the flowgraph circular it spits out bad code for the deep_dependence.stan example.
What's the bad code, is it as though it's not circular at all?

> I'm not really using it here but I added the make_circular_flowgraph function and then use it here.
Those both look good to me

> Looking at the mfp function I don't really see how adding those initials to the exits would force it to continuously loop as that function just iterates over F.successors once then does a while loop popping the workstack till it's size zero. Or am I missing something?
In the if statement at the end [here](https://github.com/SteveBronder/stanc3/blob/9db14aed32d69541e3859227854dd2a2b1cf0cdb/src/analysis_and_optimization/Monotone_framework.ml#L844), it's checking whether the most recent update made any change, and if it did, it adds all of the changed node's successors to the queue to recompute. So I would expect that when the final statements are visited and their exit sets are updated, their successors (now the initial statements) would then be recomputed with the new variable set. I'm not sure why that isn't happening

**Comment by SteveBronder** [comment]
> What's the bad code, is it as though it's not circular at all?

Yeah, so in the example model I posted above it figures out that it has to demote `tp_6` since it depends on `tp_7` but then does not go any farther up the dependency stack

> In the if statement at the end here, it's checking whether the most recent update made any change, and if it did, it adds all of the changed node's successors to the queue to recompute. So I would expect that when the final statements are visited and their exit sets are updated, their successors (now the initial statements) would then be recomputed with the new variable set. I'm not sure why that isn't happening

Oh I think I got it! So what I was doing was just adding the initials to any key in the map with an empty set as the second element. But we want to add them to the bottom of the flowgraph right? I just changed this and got it working, though it looks a little weird. Essentially in `make_circular_flowgraph` I also pass the reverse flowgraph (whose initials are the end nodes). Then I look at the keys for the forward flowgraph's map and if that key is in the initial set of the reverse flowgraph I add the forward flowgraphs initials to that map element.

It feels a little weird, but that does do what we want where it starts back at the top and then as long as nothing changes it is allowed to exit

**Comment by SteveBronder** [comment]
(also fyi the error here is from something down in Stan math I put a fix in for already)

**Comment by rybern** [comment]
@SteveBronder Great! Not sure why the empty successors wouldn't work, but that makes sense too. So working now?

**Comment by SteveBronder** [comment]
Once that math PR goes in then yes!

**Comment by wds15** [comment]
I have a bottle of red wine in my kitchen on hold for when this gets merged! Looking forward to say cheers…

File: test/integration/cli-args/stanc.t
Comment (rok-cesnovar) [inline test/integration/cli-args/stanc.t:25]: We never really had a good plan for these stanc3 flags, so we are a bit all over the place with them. 

I am not sure I have a better idea, but the `-fno-X` seems like a C++ compiler flag or is this style also common in other areas?

@WardBrian already brought up the idea of enabling individual optimization flags. Do we maybe want to have the optimization levels but then a flag to disable individual optimizatons?

Comment (rok-cesnovar) [reply test/integration/cli-args/stanc.t:25]: We can also merge this without discussing this, there is plenty of time before the release.

**Comment by rok-cesnovar** [comment]
The tests are now failing with:

```
n file included from /mnt/nvme1n1/workspace/stanc3_PR-955/performance-tests-cmdstan/cmdstan/stan/lib/stan_math/stan/math/prim/prob/lkj_corr_cholesky_log.hpp:6:
/mnt/nvme1n1/workspace/stanc3_PR-955/performance-tests-cmdstan/cmdstan/stan/lib/stan_math/stan/math/prim/prob/lkj_corr_cholesky_lpdf.hpp:24:3: error: no matching function for call to 'check_lower_triangular'
  check_lower_triangular(function, "Random variable", L_ref);
  ^~~~~~~~~~~~~~~~~~~~~~
../../test/integration/good/function-signatures/distributions/multivariate/continuous/lkj_corr_cholesky_log.hpp:266:32: note: in instantiation of function template specialization 'stan::math::lkj_corr_cholesky_lpdf<false, stan::math::var_value<Eigen::Matrix<double, -1, -1, 0, -1, -1>, void>, double>' requested here
      transformed_param_real = lkj_corr_cholesky_lpdf<false>(p_matrix,
                               ^
../../test/integration/good/function-signatures/distributions/multivariate/continuous/lkj_corr_cholesky_log.hpp:568:14: note: in instantiation of function template specialization 'lkj_corr_cholesky_log_model_namespace::lkj_corr_cholesky_log_model::log_prob_impl<false, false, Eigen::Matrix<stan::math::var_value<double, void>, -1, 1, 0, -1, 1>, Eigen::Matrix<int, -1, 1, 0, -1, 1>, nullptr, nullptr>' requested here
      return log_prob_impl<propto__, jacobian__>(params_r, params_i, pstream);
             ^
/mnt/nvme1n1/workspace/stanc3_PR-955/performance-tests-cmdstan/cmdstan/stan/src/stan/model/model_base_crtp.hpp:95:50: note: in instantiation of function template specialization 'lkj_corr_cholesky_log_model_namespace::lkj_corr_cholesky_log_model::log_prob<false, false, stan::math::var_value<double, void> >' requested here
    return static_cast<const M*>(this)->template log_prob<false, false>(theta,
                                                 ^
../../test/integration/good/function-signatures/distributions/multivariate/continuous/lkj_corr_cholesky_log.hpp:68:3: note: in instantiation of member function 'stan::model::model_base_crtp<lkj_corr_cholesky_log_model_namespace::lkj_corr_cholesky_log_model>::log_prob' requested here
  ~lkj_corr_cholesky_log_model() { }
  ^
/mnt/nvme1n1/workspace/stanc3_PR-955/performance-tests-cmdstan/cmdstan/stan/lib/stan_math/stan/math/prim/err/check_lower_triangular.hpp:27:13: note: candidate template ignored: requirement 'is_eigen<var_value<Matrix<double, -1, -1, 0, -1, -1>, void> >::value' was not satisfied [with T_y = stan::math::var_value<Eigen::Matrix<double, -1, -1, 0, -1, -1>, void>]
inline void check_lower_triangular(const char* function, const char* name,
            ^
```

**Comment by SteveBronder** [comment]
Oh whoops yes lkj is not supported by the new matrix type yet

File: test/integration/cli-args/stanc.t
Comment (SteveBronder) [reply test/integration/cli-args/stanc.t:25]: Yes so I think what we want to do is 

1. Merge #1079 which has some of the code in the PR that fixes something in the monotone framework 
2. Merge #549 so we have an O1
3. Merge this and turn it on at O1

Personally I'd prefer if we merged this immediately after the latest release. While I'm confident in it, I'd rather have a release cycle on dev just to make sure try to find any weird quirks

Comment (rok-cesnovar) [reply test/integration/cli-args/stanc.t:25]: Cool with that yeah, although if its hidden behind O1, its less of a problem.

**Comment by SteveBronder** [comment]
@rybern alright I think this is good to go! 

I currently have things set to have this turned on by default at `O0`. I think what I'd like to do is have this happen at `O1` only and merge before the release, then after the release I'd actually like to make `O1` the default option. I like this scheme because then we get all our optimizations and we also get a whole release cycle with them on in master so that if there are any bugs we have a whole release cycle to find them.

Does that work for you?

**Comment by WardBrian** [comment]
I like the idea of this / #1029 to be on in O1, but make O1 the default after the next release. I think O0 staying as a 'do absolutely nothing' option is valuable

File: test/integration/cli-args/stanc.t
Comment (WardBrian) [inline test/integration/cli-args/stanc.t:42]: This probably shouldn't be here

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: @SteveBronder I'm still wondering about this, could you double check these test cases? Suppose `b` is SoA and `c` and `d` are AoS:
* `matrix a = size(b) > 1 ? c : d;`
* `matrix a = size(b) * c;`

Do we successfully demote `a` in those cases?

**Comment by rybern** [comment]
That scheme sounds good to me @SteveBronder. +1 for @WardBrian's comment about keeping `O0`.

The last thing I want to double check on is https://github.com/stan-dev/stanc3/pull/955/files#r713252249, because that'd potentially give correctness issues

File: test/integration/cli-args/stanc.t
Comment (SteveBronder) [reply test/integration/cli-args/stanc.t]: Removed!

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: I just added these as a test and they look good!

https://github.com/stan-dev/stanc3/pull/955/files#diff-58981cfa9e197e07f3bae4252ea70ce886c138fe9b479bcbfd185fbd9f004df2

**Comment by SteveBronder** [comment]
Alrighty @rybern I just added the above tests, if everything compiles correctly on Jenkins then I'm going to remove soa from `O0` and then I think we are good to go!

File: src/analysis_and_optimization/Mem_pattern.ml
Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: Is there a foolproof way of telling what's AoS vs. SoA? Looking at the CPP for that test [here](https://github.com/stan-dev/stanc3/blob/98c62e46777a9ca284f04e374e2ca9737b324215/test/integration/good/compiler-optimizations/mem_patterns/cpp.expected#L2581) and it looks like `b_soa` might be AoS, which would affect the results. Also appears to be AoS in an MIR dump:
```
 (((pattern
     (Decl (decl_adtype AutoDiffable) (decl_id b_soa)
      (decl_type
       (Sized
        (SMatrix AoS
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
      (initialize true)))
    (meta <opaque>))
```

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: For that line you linked I'm seeing 

```
stan::conditional_var_value_t<local_scalar_t__, 
        Eigen::Matrix<local_scalar_t__, -1, -1>> b_soa
```

`stan::conditional_var_value_t` is a type trait that, if `local_scalar_t__` is `var` gives back a `var<Eigen::Matrix<local_scalar_t__, -1, -1>>` if `local_scalar_t__` is a `var` and otherwise gives back `Eigen::Matrix<local_scalar_t__, -1, -1>`. This is to handle `double` and `fvar<T>` as `local_scalar_t__`. Though what I'd like to do in the future is just have the tools around to write the `var` specialization of the `log_prob` function so we don't have to do that. 

The other one is a bug when I was printing out the transformed MIR (thanks!). I needed to print out the optimized transformed MIR. Now looks like how we expect!

```
((pattern
     (Decl (decl_adtype AutoDiffable) (decl_id b_soa)
      (decl_type
       (Sized
        (SMatrix SoA
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly))))
         ((pattern (Lit Int 10))
          (meta ((type_ UInt) (loc <opaque>) (adlevel DataOnly)))))))
      (initialize true)))
    (meta <opaque>))
```

https://github.com/stan-dev/stanc3/blob/f50fb6862b32342a2c4be50ca01aff48fc90b5b6/test/integration/good/compiler-optimizations/mem_patterns/transformed_mir.expected#L5887

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: If I comment out the `cond_aos` line, I'm getting no demotions at all, can you replicate that?

I think the `cond_aos` line is getting added to the initial AoS set because it's setting [`is_all_rhs_aos`](https://github.com/stan-dev/stanc3/blob/98c62e46777a9ca284f04e374e2ca9737b324215/src/analysis_and_optimization/Mem_pattern.ml#L344) to true, which doesn't seem right, I think the `~set` and `~subset` are switched there. Then `cond_aos` being `AoS` demotes everything else

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: Noting it here so I don't forget, it's worth checking `matrix a = transpose(c)` as well, since I'm not sure how `row(b) * c` demoted

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:387]: Pinging just in case the reply didn't, don't mean to rush you @SteveBronder

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:422]: Hey! Yes so this took me down a bit of a rabbit hole today. 

> I think the cond_aos line is getting added to the initial AoS set because it's setting is_all_rhs_aos to true, which doesn't seem right, I think the ~set and ~subset are switched there. Then cond_aos being AoS demotes everything else

Yes so I think those do need to be flipped.

I think for this line 

```stan
matrix[10, 10] int_aos_mul_aos = rows(row_soa) * empty_user_func_aos;
```

https://github.com/stan-dev/stanc3/pull/955/files#diff-3887bc12dee026edee6c3df0cde04b23cf6dd20949a84e78d04000fad60da677R27

it also didn't recognize in the initial variable sweep that `empty_user_func_aos` was AoS and so that lhs can also be demoted before even going to the monotone framework. So what I ended up doing was making the initial demotion search into a fold that keeps track of previously demoted values so then at this spot 

I added the stuff you posted above as a test [here](https://github.com/stan-dev/stanc3/pull/955/files#diff-3887bc12dee026edee6c3df0cde04b23cf6dd20949a84e78d04000fad60da677) and it's associated MIR is [here](https://github.com/stan-dev/stanc3/blob/e241be9c0ec061ca8eaa22e138025d02a8a4d4b7/test/integration/good/compiler-optimizations/mem_patterns/transformed_mir.expected#L11842). Just and fyi I try to label everything in those tests as `*_aos` or `*_soa` so if you ctrl+f for `SMatrix AoS` or `SMatrix SoA` the name in the decl should match up.

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:422]: @SteveBronder Gotcha. Thanks for adding the optimized MIR that makes it a lot easier. So it looks like we currently have issues with `empty_user_func_aos`, `int_aos_mul_aos` and `mul_two_aos`.

Any idea what's up with `empty_user_func_aos`? TBH I don't think I understand the fold fix.

It does look to me like the issue with `int_aos_mul_aos` and `mul_two_aos` is that `row_soa` is included as an SoA subexpression even though it's being used to produce a non-AD/eigen type. What if in `Mem_pattern.matrix_set` we don't recurse on non-AD/eigen expressions? Would that do it?

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: > Gotcha. Thanks for adding the optimized MIR that makes it a lot easier. So it looks like we currently have issues with empty_user_func_aos, int_aos_mul_aos and mul_two_aos.

> Any idea what's up with empty_user_func_aos? TBH I don't think I understand the fold fix.

Yes sorry I missed those I think I have it fixed now. It was that I needed to move `is_not_supported_func` out of `is_all_rhs_aos` and use a `||` instead of a `&&`. I ctrl+f'd for `_aos` and those all seem to be good now. 

> What if in Mem_pattern.matrix_set we don't recurse on non-AD/eigen expressions? Would that do it?

I think the issue was the above and so it looked like the issue was in `matrix_set` but it was actually the weird logic bits

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: @SteveBronder Awesome yes those looked fixed!

> I think the issue was the above and so it looked like the issue was in matrix_set but it was actually the weird logic bits

Looks like you made the `matrix_set` change, was that unintentional?

I noticed some odd behavior around ternary expressions, though it's not as critical because they only over-demote:
```
    // all of these demote everything including LHS and row_soa
    matrix[10, 10] ternary_soa = 1 ? row_soa : row_soa;
    matrix[10, 10] ternary_mixed_aos = 1 ? row_soa : user_func_aos;
    matrix[10, 10] ternary_mult_soa = row_soa * (1 ? row_soa : row_soa);
```

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: > Looks like you made the matrix_set change, was that unintentional?

It was intentional, looking at it I think you are right that the way I was doing that could be wrong so I changed it like you said as I think that makes more sense.

Yes for ternary expressions C++ demands that the LHS and RHS types of the ternary operator match exactly so you can't have one side returning an SoA matrix with the other returning an AoS

Comment (rybern) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: Gotcha. I see how it's hard to just demote the ternary subexpression with how it's set up. We could have it only demote when either side is AoS, but again it's not critical. In any case happy to call it.

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: ^So since they both have to match and can be confusing I forcefully demote both sides for all ternary expressions. I think at another time I want to come back to this and make these nicer

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: Yes absolutely lets rock!!

Comment (SteveBronder) [reply src/analysis_and_optimization/Mem_pattern.ml:420]: (fyi I just updated the PR to move the soa optim to only O1 so I think we should be good to go)

**Review by rybern — APPROVED** [review]
We figured out https://github.com/stan-dev/stanc3/pull/955/files#r781792609 so LGTM!

**Comment by SteveBronder** [comment]
D-O-N-E

**Comment by andrjohns** [comment]
Thanks all for sticking through this! The `varmat` future is bright!

**Comment by wds15** [comment]
WHOW!!! Can I use it now? How?? I wanna test!!

Seriously... is there a simple way to make a test drive with my favourite model? Just do -O1 as argument to the parser and that's it?

I am really looking forward to this!

**Comment by SteveBronder** [comment]
Just O1 should do it!

**Pull request merged**
